### ISSUE TITLE
malloc crash “randomly”

### ISSUE THREAD

**Issue by rastersize** [SwiftDocOrg/swift-doc#78]
Hey!

When trying to generate docs for the libraries in an app I’m working on I keep running into `malloc` crashes. This happens for different libraries for different runs. Both when using the markdown and HTML formats.

Basically I run this command for a number of modules:

```shell
swift doc generate \
  --output Libraries/$module_name \
  --module-name $module_name \
  --output artifacts/docs/API/$module_name \
  [--format html]
```

The errors I see are:

```
swift-doc(44797,0x70000879f000) malloc: Double free of object 0x7ff8b6d11c00
swift-doc(44797,0x70000879f000) malloc: *** set a breakpoint in malloc_error_break to debug
```

```
swift-doc(50645,0x700003d39000) malloc: *** error for object 0x7ff087124ba0: pointer being freed was not allocated
swift-doc(50645,0x700003d39000) malloc: *** set a breakpoint in malloc_error_break to debug
```

For now I’ve worked around it by retrying the `swift doc` operation on failures. Which seems to generate all docs.

---

Side note, I tried to build the project in Xcode to help debug the issue. Unfortunately Xcode (11.4 (11E146)) fails to build with an error:

```
Swift Compiler Error Group
/Applications/Xcode-1140.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/libxml2/libxml/module.modulemap:1:8: Redefinition of module 'libxml2'
/Users/.../swift-doc/.build/checkouts/Markup/Modules/module.modulemap:1:8: Previously defined here
```

I can file a separate issue for this if you want?

**Comment by mattt** [comment]
Thanks for reporting this, @rastersize. I've seen this come up in development as well. I'm almost certain that the root cause of this is [our implementation of `parallelMap`](https://github.com/SwiftDocOrg/swift-doc/blob/85b5872e3cdb58be273d3dd2315fa101acbfbe3c/Sources/SwiftDoc/Extensions/Array%2BParallel.swift). Unfortunately, I've also been unable to leverage Xcode's debugging stack (#60), and am left to guess at why this is randomly failing.

**Comment by kean** [comment]
It might be related to the fact that `concurrentPerform` sets the number of iterations to a number of input pages. This is not advisable to do.

**Comment by kean** [comment]
The crash is in Interface.swift:

```swift
for `extension` in extensionsByExtendedType[symbol.id.description] ?? [] {
    inheritedTypeNames.formUnion(`extension`.inheritance)
}
```

Relatively hard to repro, the problem is that `extensionsByExtendedType` is lazy and lazy isn't thread safe.

P.S. I'm not sure you need `parallelForEach` for this scenario. Writing to disk in parallel from 40 different threads doesn't sound like a great idea.

**Comment by mattt** [comment]
This should be fixed by #99. 

@rastersize I'm working to cut a new release (1.0.0-beta.3) this week, but you're welcome to try the latest in `master` in the meantime.

**Comment by rastersize** [comment]
Great news! I’ll see if I have time to try it out later this week.
